File: C++/Stacks & Queues/13_reverse_stack.cpp
/*
Recursing while popping top elements and storing in function call stack &
Inserting elements in bottom of stack.
[recursing till stack is empty (elements are stored in function call stack). inserting element and pushing rest elements]
*/

//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
public:
    // Function to insert an element at the bottom of the stack
    void insert_at_bottom(stack<int>& St, int x) {
        if (St.empty()) {
            // If the stack is empty, push the element and return
            St.push(x);
            return;
        }

        // Store the top element of the stack and pop it
        int temp = St.top();
        St.pop();

        // Recursively insert the element at the bottom
        insert_at_bottom(St, x);

        // Push back the stored top element after the recursive call
        St.push(temp);
    }

    // Function to reverse the elements of a stack
    void Reverse(stack<int>& St) {
        if (St.empty())
            return;  // If the stack is empty, return

        // Store the top element of the stack and pop it
        int num = St.top();
        St.pop();

        // Recursively reverse the remaining elements of the stack
        Reverse(St);

        // Insert the stored element at the bottom of the reversed stack
        insert_at_bottom(St, num);
    }
};

//{ Driver Code Starts.


int main() {
    int T;
    cin >> T;
    while (T--) {
        int N;
        cin >> N;
        stack<int> St;
        for (int i = 0;i < N;i++) {
            int x;
            cin >> x;
            St.push(x);
        }
        Solution ob;
        ob.Reverse(St);
        vector<int>res;
        while (St.size())
        {
            res.push_back(St.top());
            St.pop();
        }
        for (int i = res.size() - 1;i >= 0;i--)
        {
            cout << res[i] << " ";
        }

        cout << endl;
    }
}
// } Driver Code Ends

File: Python/Stacks & Queues/13_reverse_stack.py
"""
Idea:
Recursing while popping top elements and storing in function call stack &
Inserting elements in bottom of stack.
[recursing till stack is empty (elements are stored in function call stack). inserting element and pushing rest elements]
"""

# User function Template for python3

from typing import List


class Solution:
    # Function to insert an element at the bottom of the stack
    def insert_at_bottom(self, stack, x):
        if not stack:
            # If the stack is empty, append the element and return
            stack.append(x)
            return

        # Pop the top element of the stack
        temp = stack.pop()

        # Recursively insert the element at the bottom
        self.insert_at_bottom(stack, x)

        # Append back the popped element after the recursive call
        stack.append(temp)

    # Function to reverse the elements of a stack
    def reverse(self, stack):
        if not stack:
            return  # If the stack is empty, return

        # Pop the top element of the stack
        num = stack.pop()

        # Recursively reverse the remaining elements of the stack
        self.reverse(stack)

        # Insert the popped element at the bottom of the reversed stack
        self.insert_at_bottom(stack, num)


# {
# Driver Code Starts

# Initial Template for Python 3


for _ in range(int(input())):
    N = int(input())
    St = list(map(int, input().split()))
    ob = Solution()
    ob.reverse(St)
    print(*St)
# } Driver Code Ends
